# __Data Mining Assignment 1__
> Name : Ilias Dimos , Charilaos Petrakogiannis <br />
> AM : f2822102 , f2822112 <br />
> Course : Mining Big Datasets <br />
> Email : f2822102@aueb.gr , f2822112@aueb.gr <br />

# __Goal of the Assignment__
The goal of this assignment is to implement a simple workflow that will assess the similarity between bank customers and suggest for any input customer a list of his/her 10 most similar other customers. Moreover, you will be using these results to predict the rating of the customer to the bank. To calculate the similarity between customers you will first have to compute the dissimilarity for every given attribute as discussed in lecture “Measuring Data Similarity”.In order to fulfill this assignment, you will have to perform the following tasks:

## Usefull  Libraries for the Assignment

In [1]:
import pandas as pd 
pd.set_option('display.max_rows', 50)
import warnings 
warnings.simplefilter(action='ignore', category=FutureWarning)
from IPython.display import display
pd.options.mode.chained_assignment = None

## __1) Import and pre-process the dataset with customers.__

Download the bank.csv dataset from moodle. This dataset is related with direct marketing campaigns of a Portuguese banking institution. The marketing campaigns were based on phone calls to access the opinion of the customer to the bank service and products. The dataset includes `43191` bank customer profiles with `10 attributes` each. Below is a description of the available attributes:
* `Age`: The age of the customer.
* `Job`: type of job (admin, unknown, unemployed, management, housemaid, entrepreneur, student, blue-collar, self-employed, retired, technician, services).
* `Marital Status`: Married, Single, Divorced.
* `Education`: Primary, Secondary, Tertiary.
* `Default`: If the customer has credit in default (yes/no).
* `Balance`: average yearly balance, in euros.
* `Housing`: if the customer has a housing loan (yes/no).
* `Loan`: if the customer has a personal loan (yes/no)
* `Customer Rating`: The rating of the bank from the customer (Poor, Fair, Good, Very Good, Excellent).
* `Products` : An array containing the bank products (1-20) each customer has.

For any `numerical values missing`, you should replace them with the `average value` of the attribute in the dataset (rounded to the nearest integer). The replaced average values calculated should be reported in the pdf.

##### Importing the `bank.csv` file.

* In the csv file of banks dataset the row that represents the titles of our columns has an index equal to 1.
* Importing the csv file via Pandas the titles' row has no index and the index of the first customer begins from 0. 
* In order to   have a data frame that starts from a position with index equal to 1 we had to change it manualy with the function :    `banks.index=banks.index+1`.

In [2]:
banks = pd.read_csv("bank.csv",sep=";")
banks.index=banks.index+1
banks

Age           Job   Marital  Education Default  Balance Housing Loan  \
1      33.0  entrepreneur   married  secondary      no        2     yes  yes   
2      35.0    management   married   tertiary      no      231     yes   no   
3       NaN    management    single   tertiary      no      447     yes  yes   
4      42.0  entrepreneur  divorced   tertiary     yes        2     yes   no   
5      58.0       retired   married    primary      no      121     yes   no   
...     ...           ...       ...        ...     ...      ...     ...  ...   
43187  51.0    technician   married   tertiary      no      825      no   no   
43188  71.0       retired  divorced    primary      no     1729      no   no   
43189  72.0       retired   married  secondary      no     5715      no   no   
43190  57.0   blue-collar   married  secondary      no      668      no   no   
43191  37.0  entrepreneur   married  secondary      no     2971      no   no   

          Rating                   Products  
1           poor               1,3,16,17,19  
2           good                     4,8,16  
3           fair                       7,16  
4           fair       1,3,8,10,11,12,18,19  
5           good           4,5,6,7,11,18,19  
...          ...                        ...  
43187  very_good                      12,15  
43188  very_good                    4,17,19  
43189  very_good  5,7,8,9,10,11,13,16,17,18  
43190  very_good                      2,3,7  
43191  very_good           3,6,7,9,12,15,18  

[43191 rows x 10 columns]

##### Dataframe's Data Types.

In [3]:
result = banks.dtypes
print("Data Types:")
print(result)

Data Types:
Age          float64
Job           object
Marital       object
Education     object
Default       object
Balance        int64
Housing       object
Loan          object
Rating        object
Products      object
dtype: object


##### __Finding any Missing Values__
* Observing the the shell below we can see that we have 1000 missing values in `Age` column. Based on the instructions of the assignment we will fill the missing values with th average of all the ages.

In [4]:
banks.isna().sum()

Age          1000
Job             0
Marital         0
Education       0
Default         0
Balance         0
Housing         0
Loan            0
Rating          0
Products        0
dtype: int64

##### __Filling the Missing Values__
* Now that we have found the missing values we are ready to fill them using the Average Age of the whole column.
* With the describe function we can see that the average `age` of the column is approximately 41 years, so we expect our filling function to return the same number.

In [5]:
banks['Age'].describe()

count    42191.000000
mean        40.792065
std         10.540425
min         18.000000
25%         33.000000
50%         39.000000
75%         48.000000
max         95.000000
Name: Age, dtype: float64

* With the `fillna` function we managed to fill the missing values of the `Age` column.
* After that we converted the `Age` column from float to interger. We perfomed this action in this step because we werent allowed to do it without replacing the missing values first.
* We used the `round` function in order to round them to the nearest integer based on the assignment's instructions.

In [6]:
banks['Age'] = banks['Age'].fillna((banks['Age'].mean())).round()

In [7]:
banks['Age'] = banks['Age'].astype(int)

* As you can see the missing values have been replaced correclty with the average value of the `Age` column.

In [8]:
banks.isna().sum()

Age          0
Job          0
Marital      0
Education    0
Default      0
Balance      0
Housing      0
Loan         0
Rating       0
Products     0
dtype: int64

##### Final Dataset Insight

In [9]:
banks

Age           Job   Marital  Education Default  Balance Housing Loan  \
1       33  entrepreneur   married  secondary      no        2     yes  yes   
2       35    management   married   tertiary      no      231     yes   no   
3       41    management    single   tertiary      no      447     yes  yes   
4       42  entrepreneur  divorced   tertiary     yes        2     yes   no   
5       58       retired   married    primary      no      121     yes   no   
...    ...           ...       ...        ...     ...      ...     ...  ...   
43187   51    technician   married   tertiary      no      825      no   no   
43188   71       retired  divorced    primary      no     1729      no   no   
43189   72       retired   married  secondary      no     5715      no   no   
43190   57   blue-collar   married  secondary      no      668      no   no   
43191   37  entrepreneur   married  secondary      no     2971      no   no   

          Rating                   Products  
1           poor               1,3,16,17,19  
2           good                     4,8,16  
3           fair                       7,16  
4           fair       1,3,8,10,11,12,18,19  
5           good           4,5,6,7,11,18,19  
...          ...                        ...  
43187  very_good                      12,15  
43188  very_good                    4,17,19  
43189  very_good  5,7,8,9,10,11,13,16,17,18  
43190  very_good                      2,3,7  
43191  very_good           3,6,7,9,12,15,18  

[43191 rows x 10 columns]

## 2) Compute data (dis-)similarity.

To assess the similarity between the customers you could form the dissimilarity matrix for all given attributes. As described in lecture “Measuring Data Similarity”, for every given attribute you first distinguish its type (categorical, ordinal, numerical or set) and then compute the dissimilarity of its values accordingly. For set similarity use the Jaccard similarity between sets. Then, you can calculate the average of the computed dissimilarities to derive the dissimilarity over all attributes. Depending on the machine used to implement this assignment you should decide whether it is feasible to
compute the dissimilarity matrices, or, have the computations performed on-the-fly for a pair of customers.

* For us to implement the aformentioned task, firstly we had to distinguish the types of our variables:
    1. `Age` (numeric)
    2. `Job` (categorical)
    3. `Marital` (categorical)
    4. `Education` (ordinal)
    5. `Default` (categorical)
    6. `Balance` (numeric)
    7. `Housing` (categorical)
    8. `Loan` (categorical)
    9. `Rating` (ordinal)
    10. `Products` (set)
    
    


* Secondly, in order for us to proceed to the creation of the (dis-)similarity matrix, we had to convert the previous data frame in a way that would be possible to calculate the dissimilarities between its attributes:
  * For the below-mentioned variables we have replaced their `levels` with `numbers`.

##### For `education` variable

In [10]:
banks['Education'].replace({'primary': 1, 'secondary': 2, 'tertiary': 3},inplace=True)

##### For `Rating` variable

In [11]:
banks['Rating'].replace({'poor': 1, 'fair': 2, 'good': 3, 'very_good':4, 'excelent':5},inplace=True)

##### For `Products` variable
* In this case we used the `.split(",")` command in order to separate the attributes in the `Product` column , otherwise the system by default recognises them as a single string.

In [12]:
%%time
for i in range(1,43192):
    banks['Products'][i]=banks['Products'][i].split(",")

Wall time: 35 s


* Below, we can see that now the columns `'Education'` and `'Rating'` contain numbers instead of characters and the column `Products` has been transformed into a list that contains the unique products of each customer. 
* These transformations gives us the opportunity to conduct calculations between ordinal attributes.

##### Banks insight after the tranformations

In [13]:
banks

Age           Job   Marital  Education Default  Balance Housing Loan  \
1       33  entrepreneur   married          2      no        2     yes  yes   
2       35    management   married          3      no      231     yes   no   
3       41    management    single          3      no      447     yes  yes   
4       42  entrepreneur  divorced          3     yes        2     yes   no   
5       58       retired   married          1      no      121     yes   no   
...    ...           ...       ...        ...     ...      ...     ...  ...   
43187   51    technician   married          3      no      825      no   no   
43188   71       retired  divorced          1      no     1729      no   no   
43189   72       retired   married          2      no     5715      no   no   
43190   57   blue-collar   married          2      no      668      no   no   
43191   37  entrepreneur   married          2      no     2971      no   no   

       Rating                              Products  
1           1                    [1, 3, 16, 17, 19]  
2           3                            [4, 8, 16]  
3           2                               [7, 16]  
4           2         [1, 3, 8, 10, 11, 12, 18, 19]  
5           3              [4, 5, 6, 7, 11, 18, 19]  
...       ...                                   ...  
43187       4                              [12, 15]  
43188       4                           [4, 17, 19]  
43189       4  [5, 7, 8, 9, 10, 11, 13, 16, 17, 18]  
43190       4                             [2, 3, 7]  
43191       4              [3, 6, 7, 9, 12, 15, 18]  

[43191 rows x 10 columns]

* Moreover, we've changed the data types, because now `Education` and `Rating` contain integers, not strings!

##### Tranformation from `Object` to `Numeric` for `Education` variable

In [14]:
print(pd.to_numeric(banks.Education))

1        2
2        3
3        3
4        3
5        1
        ..
43187    3
43188    1
43189    2
43190    2
43191    2
Name: Education, Length: 43191, dtype: int64


##### Tranformation from `Object` to `Numeric` for `Rating` variable

In [15]:
print(pd.to_numeric(banks.Rating))

1        1
2        3
3        2
4        2
5        3
        ..
43187    4
43188    4
43189    4
43190    4
43191    4
Name: Rating, Length: 43191, dtype: int64


##### Desired data types for the continuation of the assignment

In [16]:
result = banks.dtypes
print("Data Types:")
print(result)

Data Types:
Age           int32
Job          object
Marital      object
Education     int64
Default      object
Balance       int64
Housing      object
Loan         object
Rating        int64
Products     object
dtype: object


* Because is imposible to create a `43191x43191 ( 1,865,462,481 memory slots ) ` matrix due to the lack of memory we will implement the computation of the     dissimilarity on the fly for a pair of customers.In the next section of the assignment we created 10 lists , each one of   them corresponds to a specific customer. The list will be filled with the average `Similarities`  between the given customer and the rest of them. So at the end of the day will have filled 10 lists with 43191 average similarities. 
* In order to have a `verification measure` in our code we will compute the similarity and for the customer with his self. If all the computations of the similarities are correct the average similarity for the customer himself should be 1.

## 3) Nearest Neighbor (NN) search

Using the implementation of the previous step, you will calculate the 10-NN (most
similar) customers for the customers with ids listed below (customer id=line number-
1):
`1200, 3650, 10400, 14930, 22330, 25671, 29311, 34650, 39200, 42000`.
For this task your script must take as input the customer-id and `return the list` of her
10 nearest neighbors (most similar), along with the corresponding similarity score.

##### Creation of the lists with the most similar customers for the given customers.
In the code block below we create 10 lists for our given customers.With the first __`for loop`__ we were able to run through them and with the second __`for loop`__ we managed to compute the `dissimilarity` of each one of them with the other 43191 customers.
Because of the mixed data we have in our dataset we had to compute different `dissimilarities` for each type of variable. At the end of the day the 4 different dissimilarities we have computed are the ones below:

   * For the `Nominal` variables we used : $$ d(a,b) = 1~if~a≠b,~0~otherwise $$
    
   * For the `ordinal` variables we used : $$ d(a,b)=\frac{|rank(a)-rank(b)|}{maxRank - minRank} $$
    
   * For the `numeric` variables we used : $$ d(a,b)=\frac{|a-b|}{maxValue - minValue} $$
    
   * For the `set` variables we used :     $$ Jaccard(S1,S2) =1 - \frac{|S1\cap S2|}{|S1\cup S2|} $$  

* The average `Similarity` of the customers is calculated by substructing the 1 from the already computed disimilarities which are stored in the `score` variable.
* At the end of the __`for loop`__ we used the `.append` function in order to fill our lists for each given customer's similarity with the rest of the customers.
* We have also added to our given customers the customer with id = 1 in order to verify our results with the ones given from the assignment's instructions.

* The `Run Time` for the below mentioned procudure is approximately 2 minutes and 30 seconds.

In [17]:
%%time
cust_ids=[1,1200, 3650, 10400, 14930, 22330, 25671, 29311, 34650, 39200, 42000]

similarity_list_1  = []
similarity_list_1200  = []
similarity_list_3650  = []
similarity_list_10400 = []
similarity_list_14930 = []
similarity_list_22330 = []
similarity_list_25671 = []
similarity_list_29311 = []
similarity_list_34650 = []
similarity_list_39200 = []
similarity_list_42000 = []

for c in cust_ids:
   
    for i in range(1,43192):
        age =  abs( banks.Age[c] - banks.Age[i] ) / ( banks.Age.max() - banks.Age.min() )
        job =  [0 if banks.Job[c] == banks.Job[i] else 1]
        marital =   [0 if banks.Marital[c] == banks.Marital[i] else 1]
        education =   abs( banks.Education[c] - banks.Education[i]) / 2 # maxrank - minrank = 3 - 1 = 2 
        default = [0 if banks.Default[c] == banks.Default[i] else 1]
        balance =  abs( banks.Balance[c] - banks.Balance[i] )/ ( banks.Balance.max() - banks.Balance.min() )
        housing =  [0 if banks.Housing[c] == banks.Housing[i] else 1]
        loan =  [0 if banks.Loan[c] == banks.Loan[i] else 1]
        rating =  abs( banks.Rating[c] - banks.Rating[i]) / 4  # maxrank - minrank = 5 - 1 = 4
        
        # Jaccard Dissimilarity
        list1= set(banks.Products[c]) 
        list2= set(banks.Products[i]) 
        
        intersection = len(list1.intersection(list2)) 
        union = len(list1.union(list2)) 
        products = 1 - float(intersection / union)

        score = 1-( age + job + marital + education + default + balance + housing + loan +  rating + products ) / 10
        
        if   c==1200:
            similarity_list_1200.append(score)
        elif c==3650:
            similarity_list_3650.append(score)
        elif c==10400 : 
            similarity_list_10400.append(score)
        elif c==14930 : 
            similarity_list_14930.append(score)
        elif c== 22330 : 
            similarity_list_22330.append(score)
        elif c== 25671 : 
            similarity_list_25671.append(score)
        elif c== 29311 : 
            similarity_list_29311.append(score)
        elif c== 34650 : 
            similarity_list_34650.append(score)
        elif c== 39200 : 
            similarity_list_39200.append(score)
        elif c== 42000 : 
            similarity_list_42000.append(score)
        elif c== 1 : 
            similarity_list_1.append(score)

Wall time: 2min 51s


* In this part of the assignment is very important to mention that the above procedure would have been implemented much faster with the use of `dictionaries` instead of `lists`, but we implemented the question based on the assignment's instructions.

##### Changing the Lists into Dataframes
* In this step of the code we used the pandas library in order to make our lists dataframes for each given customer. 
* We have set as `column` name : `Similarity Score ( for each given customer )` and as `Index` name : `Customer ID`
* Last but not least in order to find the 10 most similar customers we have sorted every dataframe in descenting order. In this way we were able to find the 10 highest `Similarities` among the given customer and the rest of them.
* Its very important to mention that the similarity score among every given customer and its self is 1. It can be displayed by deleting the `.iloc[1:12]` function.
* As we mention above about the similarity of the customer himself in order for our outcomes to match the ones from the assignment's instructions we use the `.iloc[1:12]` function and we print the only the 10 most similar customers.
* Also it was necessery to increase the index of the Data Frames by one because by default the Pandas set the index equal to 0.

In [18]:
df_1 = pd.DataFrame(similarity_list_1,columns=['Similarity Score'])
df_1.index=df_1.index+1
df_1.index.names = ['Customer ID']
df_1 = df_1.sort_values(by=['Similarity Score'],ascending=False).head(11).iloc[1:12]

df_1200 = pd.DataFrame(similarity_list_1200,columns=['Similarity Score'])
df_1200.index=df_1200.index+1
df_1200.index.names = ['Customer ID']
df_1200 = df_1200.sort_values(by=['Similarity Score'],ascending=False).head(11).iloc[1:12]

df_3650 = pd.DataFrame(similarity_list_3650,columns=['Similarity Score'])
df_3650.index=df_3650.index+1
df_3650.index.names = ['Customer ID']
df_3650 = df_3650.sort_values(by=['Similarity Score'],ascending=False).head(11).iloc[1:12]

df_10400 = pd.DataFrame(similarity_list_10400,columns=['Similarity Score'])
df_10400.index=df_10400.index+1
df_10400.index.names = ['Customer ID']
df_10400 = df_10400.sort_values(by=['Similarity Score'],ascending=False).head(11).iloc[1:12]

df_14930 = pd.DataFrame(similarity_list_14930,columns=['Similarity Score'])
df_14930.index=df_14930.index+1
df_14930.index.names = ['Customer ID']
df_14930 = df_14930.sort_values(by=['Similarity Score'],ascending=False).head(11).iloc[1:12]

df_22330 = pd.DataFrame(similarity_list_22330,columns=['Similarity Score'])
df_22330.index=df_22330.index+1
df_22330.index.names = ['Customer ID']
df_22330 = df_22330.sort_values(by=['Similarity Score'],ascending=False).head(11).iloc[1:12]

df_25671 = pd.DataFrame(similarity_list_25671,columns=['Similarity Score'])
df_25671.index=df_25671.index+1
df_25671.index.names = ['Customer ID']
df_25671 = df_25671.sort_values(by=['Similarity Score'],ascending=False).head(11).iloc[1:12]

df_29311 = pd.DataFrame(similarity_list_29311,columns=['Similarity Score'])
df_29311.index=df_29311.index+1
df_29311.index.names = ['Customer ID']
df_29311 = df_29311.sort_values(by=['Similarity Score'],ascending=False).head(11).iloc[1:12]

df_34650 = pd.DataFrame(similarity_list_34650,columns=['Similarity Score'])
df_34650.index=df_34650.index+1
df_34650.index.names = ['Customer ID']
df_34650 = df_34650.sort_values(by=['Similarity Score'],ascending=False).head(11).iloc[1:12]

df_39200 = pd.DataFrame(similarity_list_39200,columns=['Similarity Score'])
df_39200.index=df_39200.index+1
df_39200.index.names = ['Customer ID']
df_39200 = df_39200.sort_values(by=['Similarity Score'],ascending=False).head(11).iloc[1:12]

df_42000 = pd.DataFrame(similarity_list_42000,columns=['Similarity Score'])
df_42000.index=df_42000.index+1
df_42000.index.names = ['Customer ID']
df_42000 = df_42000.sort_values(by=['Similarity Score'],ascending=False).head(11).iloc[1:12]

* After the creation of the `Pandas Dataframes` for each customer ,  we assigned them into a variable `df_lists` and with the `for` loop we managed to `display` them in the same code block.


In [19]:
df_list = [df_1,df_1200,df_3650,df_10400,df_14930,df_22330,df_25671,df_29311,df_34650,df_39200,df_42000]

titles =[' 10 NN for Customer 1 ',' 10 NN for Customer 1200 ',' 10 NN for Customer 3650 ',
         ' 10 NN for Customer 10400 ',' 10 NN for Customer 14930 ',
         ' 10 NN for Customer 22330 ',' 10 NN for Customer 25671 ',
         ' 10 NN for Customer 29311 ',' 10 NN for Customer 34650 ',
         ' 10 NN for Customer 39200',' 10 NN for Customer 42000 ']
class color:
   BOLD = '\033[1m'
   END = '\033[0m'

for i , j   in zip(df_list,titles):
    print("----------------------------")
    print(color.BOLD + j + color.END)
    display(i)

----------------------------
 10 NN for Customer 1 


Similarity Score
Customer ID                  
16641                0.940258
12329                0.935793
1247                 0.933332
33282                0.923375
25849                0.920261
24715                0.918505
6001                 0.916558
31996                0.915727
5914                 0.915240
7894                 0.912874

----------------------------
 10 NN for Customer 1200 


Similarity Score
Customer ID                  
41726                0.954603
41664                0.926438
23961                0.923867
16983                0.923040
8756                 0.920661
4308                 0.920348
16156                0.919752
1106                 0.917888
13603                0.917460
9816                 0.915249

----------------------------
 10 NN for Customer 3650 


Similarity Score
Customer ID                  
6189                 0.950560
28141                0.937784
8067                 0.936680
6018                 0.936171
5616                 0.927949
7307                 0.923025
20262                0.918398
33475                0.916447
23347                0.913418
37973                0.913042

----------------------------
 10 NN for Customer 10400 


Similarity Score
Customer ID                  
17598                0.964383
20960                0.925819
26167                0.918353
8653                 0.915153
28056                0.914227
10516                0.913437
9097                 0.911474
15891                0.909768
11310                0.909424
23687                0.897207

----------------------------
 10 NN for Customer 14930 


Similarity Score
Customer ID                  
33854                0.957549
33177                0.952029
2706                 0.938542
8231                 0.933491
238                  0.932029
31461                0.931961
4047                 0.930308
36760                0.929137
4789                 0.928226
28025                0.928155

----------------------------
 10 NN for Customer 22330 


Similarity Score
Customer ID                  
9840                 0.931193
17137                0.915372
22789                0.915206
687                  0.910131
22250                0.907943
13617                0.900969
21326                0.895619
10589                0.887767
22615                0.884087
22884                0.879431

----------------------------
 10 NN for Customer 25671 


Similarity Score
Customer ID                  
17207                0.961472
38036                0.935065
23514                0.932468
22337                0.930909
17777                0.922078
27992                0.913298
26599                0.912098
28054                0.912002
41849                0.910125
40432                0.909962

----------------------------
 10 NN for Customer 29311 


Similarity Score
Customer ID                  
5182                 0.848625
24620                0.835948
23960                0.833641
34477                0.826136
34631                0.823541
3146                 0.822697
13991                0.822025
7172                 0.818546
6765                 0.818267
25208                0.817017

----------------------------
 10 NN for Customer 34650 


Similarity Score
Customer ID                  
6902                 0.931291
27643                0.927901
20567                0.924065
7658                 0.923906
30715                0.922335
33302                0.921600
31572                0.921122
32193                0.920330
24952                0.920001
36355                0.918987

----------------------------
 10 NN for Customer 39200


Similarity Score
Customer ID                  
13171                0.958222
29359                0.929250
37141                0.928357
13047                0.922020
9744                 0.921712
29463                0.921251
27083                0.920309
27579                0.917305
42809                0.912519
29055                0.912070

----------------------------
 10 NN for Customer 42000 


Similarity Score
Customer ID                  
32471                0.958657
28027                0.950954
31801                0.943429
41316                0.942950
39518                0.937329
22742                0.935395
41144                0.935216
42419                0.935024
32493                0.933916
39425                0.933633

## 4) Customer rating prediction

### 4.1) Calculation of the Similarities
Calculate the similarities between the given customer and all other customers and compute his 10-nn (most similar) customers. __IMPORTANT : In the similarity calculations for this step you need to exclude the customer rating attribute.__

##### Preperatory Steps for the Prediction of the `Rating` .

* Firstly, with the `.copy`  function we were able to take all the customers of our initial dataset without having to modify them again as we did in the first `Question` and store them into a new variable called `banks_2`.

* Secondly , with the `.drop` function we were able to remove the `Rating` column as we want to predict her later.

Now our dataset contains 43191 rows ( starting from the 1 ) and 9 columns.



In [20]:
banks_2 = banks.copy()
banks_2 = banks_2.drop('Rating',axis=1)
banks_2.head()

Age           Job   Marital  Education Default  Balance Housing Loan  \
1   33  entrepreneur   married          2      no        2     yes  yes   
2   35    management   married          3      no      231     yes   no   
3   41    management    single          3      no      447     yes  yes   
4   42  entrepreneur  divorced          3     yes        2     yes   no   
5   58       retired   married          1      no      121     yes   no   

                        Products  
1             [1, 3, 16, 17, 19]  
2                     [4, 8, 16]  
3                        [7, 16]  
4  [1, 3, 8, 10, 11, 12, 18, 19]  
5       [4, 5, 6, 7, 11, 18, 19]

##### Creation of the lists with the most similar customers for the first 50 customers among all the other customers of the dataset `banks_2`.
In this part of the assignment we implemented the same procedure we did in the 3rd Question. The only diffrence is that we created a list for the every single one of the first 50 customers of our new dataset with the function `customer = [[] for _ in range(n)]`  to save our `similarity` scores for every customer with the others.

At the end of the function we store our similarity scores into the customer list with the use of the `.append()` function. The specific lists contain all the similarities for the first 50 customers with all the others.

The `RunTime` of the procedure below is approximately 12 minutes. 

In [21]:
%%time

customer = [[] for _ in range(51)]

for cust in range(1,51):
       
    for i in range(1,43192):
        
        age = abs( banks_2.Age[cust] - banks_2.Age[i] ) / ( banks_2.Age.max() - banks_2.Age.min() )
        job = [0 if banks_2.Job[cust] == banks_2.Job[i] else 1]
        marital = [0 if banks_2.Marital[cust] == banks_2.Marital[i] else 1]
        education = abs( banks_2.Education[cust] - banks_2.Education[i]) / 2 # maxrank - minrank = 3 - 1 = 2 
        default = [0 if banks_2.Default[cust] == banks_2.Default[i] else 1]
        balance = abs( banks_2.Balance[cust] - banks_2.Balance[i] ) / ( banks_2.Balance.max() - banks_2.Balance.min() )
        housing = [0 if banks_2.Housing[cust] == banks_2.Housing[i] else 1]
        loan =  [0 if banks_2.Loan[cust] == banks_2.Loan[i] else 1]
        
        # Jaccard Dissimilarity
        list1= set(banks_2.Products[cust]) 
        list2= set(banks_2.Products[i]) 

        intersection = len(list1.intersection(list2)) 
        union = len(list1.union(list2)) 
        products = 1 - float(intersection / union)

        score = 1 - ( age + job + marital + education + default + balance + housing + loan  + products ) / 9
        
        customer[cust].append(score)

Wall time: 12min 30s


##### Storing the lists as dataframes using the `global` function. 

* After filling the lists with the Similarity scores we made a repeated procedure in order to make each one of the lists a dataframe. Because the lists are under the scope of the customer variable we had to use the `globals()[f"df_{i}"]` function to implement it. More precisely , now the new datasets can be called by typing the __`df_i`__ where `i` is a range from 1 to 50 corresponding to our first `50 customers`.
* Additionaly , we have sorted the values in descenting order by ` Similatiry Score` column. Because the first element of  the sorted list will be the score of the customer with its self ( and it is 1 ) we print the rows for the 1 customer to 12 using the `loc` function.
* Because the new dataframe we create starts from 0 we need to make it start from 1. Thats why it is nessecerry to increase our `Customer ID` by 1 . 
* We implement the same procedure and for the `index` in order to follow the numbering of the `Customer ID` column.

In [22]:
for i in range(1,51):
    globals()[f"df_{i}"] = pd.DataFrame(customer[i]).rename_axis(index="Customer ID").set_axis(['Similarity Score'],axis=1).sort_values(by='Similarity Score',ascending=False).reset_index().head(11).loc[1:12]
    globals()[f"df_{i}"]['Customer ID']=globals()[f"df_{i}"]['Customer ID']+1
    globals()[f"df_{i}"].index =globals()[f"df_{i}"].index+1

###### Example of the Dataframe creation

In [23]:
df_1

Customer ID  Similarity Score
2         25849          0.939179
3         16641          0.933620
4         12329          0.928658
5          1247          0.925924
6         24633          0.916667
7         33282          0.914861
8         34865          0.913619
9         32932          0.912137
10        24715          0.909450
11        12951          0.908824

* After the contruction of the DataFrames we created to each one of them a new column named `Rating`.

In [24]:
for i in range(1,51):
    globals()[f"df_{i}"]["Rating"] = " "

* Below we reset the index of the initial `Banks` dataset and we rename it as `Customer ID` in order for us to work with columns and not indexes.

In [25]:
banks.reset_index(inplace=True)
banks.index=banks.index+1
banks.rename(columns={'index':'Customer ID'},inplace=True)

In [26]:
banks

Customer ID  Age           Job   Marital  Education Default  Balance  \
1                1   33  entrepreneur   married          2      no        2   
2                2   35    management   married          3      no      231   
3                3   41    management    single          3      no      447   
4                4   42  entrepreneur  divorced          3     yes        2   
5                5   58       retired   married          1      no      121   
...            ...  ...           ...       ...        ...     ...      ...   
43187        43187   51    technician   married          3      no      825   
43188        43188   71       retired  divorced          1      no     1729   
43189        43189   72       retired   married          2      no     5715   
43190        43190   57   blue-collar   married          2      no      668   
43191        43191   37  entrepreneur   married          2      no     2971   

      Housing Loan  Rating                              Products  
1         yes  yes       1                    [1, 3, 16, 17, 19]  
2         yes   no       3                            [4, 8, 16]  
3         yes  yes       2                               [7, 16]  
4         yes   no       2         [1, 3, 8, 10, 11, 12, 18, 19]  
5         yes   no       3              [4, 5, 6, 7, 11, 18, 19]  
...       ...  ...     ...                                   ...  
43187      no   no       4                              [12, 15]  
43188      no   no       4                           [4, 17, 19]  
43189      no   no       4  [5, 7, 8, 9, 10, 11, 13, 16, 17, 18]  
43190      no   no       4                             [2, 3, 7]  
43191      no   no       4              [3, 6, 7, 9, 12, 15, 18]  

[43191 rows x 11 columns]

##### Filling the `Rating` column.

Based on the assignment's instructions we need to fill the column `Rating` based on the initial dataset `banks`. 
The multiple repeated procedure in the code block below works as follows : 
* 1) The first `for loop` is used in order to take in every iteration the dataset `i` which is stored in the `globals` function.
* 2) The second `for loop` is used to take the `j` row for the each dataset stored in `globals`.
* 3) The third `for loop` is used to iterate among the rows of the initial dataset `banks`.
* 4) The last `if statement` tests where the `j` row of the `Customers ID` column from the `i` dataset in `globals` is equal with the `k` row of the `Customer ID` from the initial dataset. Where the statement is `True` then stores in the `Rating` column the `Rating` score from the `banks` dataset to the specific customer.
* 5) The `Run Time` for the code block is approximately `2 minutes`.

In [27]:
%%time
for i in range(1,51):
    for j in range(2,12):
        for k in range(1,43192):
            if globals()[f"df_{i}"]['Customer ID'][j] == banks['Customer ID'][k]:
                globals()[f"df_{i}"]['Rating'][j] = banks['Rating'][k]

Wall time: 2min 44s


##### Displaying the Datasets for the 50 customers.

* Its important to mention that we choose to print only the customers from the second row to the 11th row of each dataset. Thats why because the dataframes are sorted in a descending way the and the first customer will be always the customer himself with Similarity Score equal to 1.

In [28]:
class color:
   BOLD = '\033[1m'
   END = '\033[0m'

for i in range(1,51):
    print('----------------------------------')
    print(color.BOLD + '10 NN for Customer : ' + color.END,i)
    display(globals()[f"df_{i}"])

----------------------------------
10 NN for Customer :  1


Customer ID  Similarity Score Rating
2         25849          0.939179      2
3         16641          0.933620      1
4         12329          0.928658      1
5          1247          0.925924      1
6         24633          0.916667      2
7         33282          0.914861      1
8         34865          0.913619      2
9         32932          0.912137      2
10        24715          0.909450      1
11        12951          0.908824      2

----------------------------------
10 NN for Customer :  2


Customer ID  Similarity Score Rating
2         14688          0.959276      3
3         31408          0.952457      3
4         11418          0.949416      3
5         35077          0.944031      3
6         37778          0.942900      3
7          9901          0.942683      2
8         33676          0.941355      3
9         30741          0.939599      3
10        25212          0.937194      3
11        31417          0.936470      3

----------------------------------
10 NN for Customer :  3


Customer ID  Similarity Score Rating
2          5707          0.943419      2
3         18687          0.936779      1
4         28014          0.925771      2
5          5578          0.925529      2
6         28453          0.916032      2
7         10346          0.914826      2
8         30525          0.914452      1
9         36309          0.911045      1
10        40194          0.909314      2
11         4297          0.908777      2

----------------------------------
10 NN for Customer :  4


Customer ID  Similarity Score Rating
2         24757          0.894851      2
3         34488          0.833887      3
4          8166          0.828797      3
5         37585          0.827414      3
6          1090          0.822212      2
7          2332          0.815783      2
8         34605          0.808728      2
9         26537          0.807451      3
10         2358          0.805193      2
11        35747          0.803750      2

----------------------------------
10 NN for Customer :  5


Customer ID  Similarity Score Rating
2          2961          0.930433      3
3          8343          0.928891      3
4          3093          0.928862      3
5         10803          0.927828      3
6         30157          0.927546      3
7         12544          0.927500      3
8         28518          0.921044      2
9          1156          0.919212      2
10           11          0.919083      3
11        21575          0.918207      3

----------------------------------
10 NN for Customer :  6


Customer ID  Similarity Score Rating
2          3452          0.944263      3
3         21662          0.940981      3
4         34296          0.940389      3
5          6064          0.939787      2
6          4707          0.938974      3
7          6000          0.937707      3
8         34350          0.937648      3
9         10562          0.935172      3
10        32406          0.933816      3
11        13950          0.933064      3

----------------------------------
10 NN for Customer :  7


Customer ID  Similarity Score Rating
2         26106          0.947986      3
3           893          0.941439      3
4         33285          0.941281      3
5         14301          0.941239      3
6          7598          0.938889      3
7         32181          0.938349      3
8         39357          0.936457      3
9         34173          0.934090      3
10         2282          0.933207      2
11        40758          0.932842      3

----------------------------------
10 NN for Customer :  8


Customer ID  Similarity Score Rating
2         35682          0.996946      3
3         42897          0.992435      3
4         13563          0.973299      2
5         10971          0.944099      3
6         12563          0.944014      2
7          8222          0.941268      3
8          8112          0.940900      2
9          8093          0.940350      3
10        37693          0.940022      3
11        37104          0.938248      3

----------------------------------
10 NN for Customer :  9


Customer ID  Similarity Score Rating
2         15759          0.942562      2
3          6423          0.939864      3
4         37494          0.935252      3
5         31689          0.933937      2
6         28707          0.933281      3
7         28978          0.932914      2
8          2667          0.932614      3
9          7570          0.932311      3
10        15729          0.931479      3
11        28385          0.930758      3

----------------------------------
10 NN for Customer :  10


Customer ID  Similarity Score Rating
2          1240          0.999980      3
3          8344          0.998624      2
4         34289          0.998431      3
5          6547          0.995191      2
6         24709          0.993864      3
7         13154          0.991135      3
8         39541          0.989858      3
9         14215          0.989776      3
10         6554          0.989760      3
11          219          0.983466      3

----------------------------------
10 NN for Customer :  11


Customer ID  Similarity Score Rating
2         15570          0.935621      3
3         21575          0.932469      3
4         28492          0.928139      3
5         10803          0.927763      3
6         18371          0.920404      2
7          3838          0.919159      3
8             5          0.919083      3
9         31616          0.918060      3
10        16500          0.917149      3
11         8237          0.916864      2

----------------------------------
10 NN for Customer :  12


Customer ID  Similarity Score Rating
2           698          0.959567      2
3          2870          0.955194      3
4          8264          0.954597      3
5         31331          0.954352      3
6         13240          0.950294      3
7         37032          0.946247      3
8         35556          0.942567      3
9          4553          0.941908      3
10        29595          0.941653      3
11        31339          0.941217      3

----------------------------------
10 NN for Customer :  13


Customer ID  Similarity Score Rating
2          4090          0.943679      3
3         22636          0.923106      3
4          7286          0.913246      3
5         37647          0.906499      3
6         14951          0.905057      3
7         22112          0.903211      2
8          3687          0.902376      3
9         33232          0.902306      3
10         3188          0.902081      3
11            5          0.901814      3

----------------------------------
10 NN for Customer :  14


Customer ID  Similarity Score Rating
2         31997          0.955285      3
3           726          0.952252      3
4           764          0.949728      2
5         17506          0.948488      3
6         26883          0.944081      3
7         36786          0.943815      3
8         32179          0.942589      3
9         28147          0.942584      3
10        26724          0.941217      3
11        15262          0.940970      3

----------------------------------
10 NN for Customer :  15


Customer ID  Similarity Score Rating
2         30554          0.946783      2
3          6589          0.940049      1
4           867          0.938892      2
5         30362          0.938665      1
6          8392          0.938056      2
7          4955          0.936811      2
8          5877          0.932075      2
9         35123          0.931573      2
10        10968          0.929942      2
11         7679          0.928644      2

----------------------------------
10 NN for Customer :  16


Customer ID  Similarity Score Rating
2         15643          0.943519      3
3         34572          0.942941      3
4         28754          0.939432      3
5         31620          0.939412      3
6         13387          0.938591      3
7         25137          0.935819      3
8         16689          0.934641      3
9          7011          0.934062      3
10         7405          0.932551      2
11        22958          0.931874      2

----------------------------------
10 NN for Customer :  17


Customer ID  Similarity Score Rating
2          2596          0.927233      2
3         33469          0.922217      2
4         29100          0.921252      2
5         36009          0.916667      2
6         25572          0.912857      2
7           943          0.908906      2
8         14845          0.908202      1
9         24357          0.907856      1
10         5908          0.907407      2
11        15159          0.906342      1

----------------------------------
10 NN for Customer :  18


Customer ID  Similarity Score Rating
2         34394          0.957742      2
3         37147          0.952602      3
4         14510          0.945301      2
5          7961          0.940540      3
6         15028          0.939613      2
7         30303          0.938558      3
8          8495          0.937887      3
9         29110          0.937790      3
10         5187          0.936889      3
11        17498          0.936652      2

----------------------------------
10 NN for Customer :  19


Customer ID  Similarity Score Rating
2          4614          0.905875      2
3         15468          0.904121      2
4         29539          0.901459      2
5         25579          0.897674      2
6         19770          0.897547      1
7         17606          0.894537      1
8         19845          0.893400      2
9         22159          0.893115      1
10        36771          0.890158      2
11         7739          0.888732      1

----------------------------------
10 NN for Customer :  20


Customer ID  Similarity Score Rating
2         28156          0.952005      2
3         23432          0.951139      3
4         13391          0.942153      3
5         30586          0.941102      3
6         15422          0.939939      3
7         39581          0.939635      3
8         36555          0.938787      3
9          2851          0.938246      3
10        24211          0.937576      2
11        30646          0.936407      3

----------------------------------
10 NN for Customer :  21


Customer ID  Similarity Score Rating
2         18665          0.960474      3
3         24121          0.944187      2
4          5273          0.943909      3
5         41238          0.942744      2
6           838          0.942424      3
7          1273          0.939577      3
8         32225          0.937125      3
9         35671          0.937049      3
10        21109          0.936593      3
11        24762          0.932643      2

----------------------------------
10 NN for Customer :  22


Customer ID  Similarity Score Rating
2           463          0.921542      2
3         30884          0.909554      1
4         16261          0.902035      2
5         21397          0.900363      2
6         23165          0.899778      1
7          6001          0.898276      1
8          7934          0.897433      1
9         16641          0.895990      1
10        35940          0.895743      2
11        16011          0.895519      2

----------------------------------
10 NN for Customer :  23


Customer ID  Similarity Score Rating
2         26204          0.942426      3
3         27351          0.921287      3
4         32183          0.913423      3
5         32930          0.912137      2
6          7604          0.911051      3
7         20933          0.907995      2
8         23532          0.907305      3
9          5668          0.906819      3
10        22313          0.906204      3
11        41725          0.905522      3

----------------------------------
10 NN for Customer :  24


Customer ID  Similarity Score Rating
2          2211          0.945253      1
3         16400          0.938540      2
4          3983          0.936962      1
5         33849          0.935616      2
6          7861          0.934864      2
7         30715          0.933209      2
8         19815          0.925131      2
9         40998          0.924483      2
10        30803          0.923477      2
11        20567          0.922519      2

----------------------------------
10 NN for Customer :  25


Customer ID  Similarity Score Rating
2         19464          0.919967      2
3         13621          0.918146      3
4         21291          0.913664      3
5         12592          0.913536      2
6         25953          0.911723      3
7         22013          0.909331      3
8         21799          0.907780      3
9         27968          0.906768      3
10        11623          0.906166      2
11        13014          0.901775      2

----------------------------------
10 NN for Customer :  26


Customer ID  Similarity Score Rating
2         34734          0.983625      2
3         40545          0.978181      3
4          2722          0.965445      4
5         11463          0.953223      3
6         33298          0.948983      2
7          5803          0.944118      3
8         31266          0.942971      3
9         36332          0.942928      3
10        16361          0.939494      2
11         7417          0.937900      3

----------------------------------
10 NN for Customer :  27


Customer ID  Similarity Score Rating
2         32995          0.957937      2
3          3573          0.935950      2
4         37053          0.926924      2
5         24252          0.926510      2
6         16279          0.925648      2
7         23255          0.924771      2
8         13452          0.922251      2
9          8806          0.921602      2
10        15889          0.919830      2
11        32371          0.919674      2

----------------------------------
10 NN for Customer :  28


Customer ID  Similarity Score Rating
2         20061          0.938987      3
3          2474          0.936623      3
4          2298          0.934194      3
5         34058          0.931431      3
6           625          0.929988      3
7         30862          0.929910      3
8         35572          0.929894      3
9         34449          0.929718      3
10        30717          0.926818      3
11        15670          0.926305      3

----------------------------------
10 NN for Customer :  29


Customer ID  Similarity Score Rating
2         25658          0.941926      3
3          4143          0.938432      3
4         17810          0.935243      3
5         11263          0.933657      3
6         23283          0.930889      3
7          4095          0.927711      3
8         22210          0.927491      3
9         23870          0.926669      3
10        24679          0.925563      4
11         7421          0.925414      2

----------------------------------
10 NN for Customer :  30


Customer ID  Similarity Score Rating
2         18399          0.952850      3
3         24844          0.928523      3
4         24450          0.925872      3
5         11675          0.921403      3
6          5926          0.921300      3
7         34814          0.916630      3
8         16435          0.914729      3
9          2965          0.913598      3
10        24042          0.913040      3
11        42314          0.912741      3

----------------------------------
10 NN for Customer :  31


Customer ID  Similarity Score Rating
2         29554          0.945447      2
3         35505          0.942083      3
4         25234          0.941731      3
5          1477          0.939356      3
6         32391          0.938741      3
7         16091          0.938205      3
8         30506          0.937044      3
9          5572          0.937034      3
10        34152          0.937027      3
11        31389          0.936614      2

----------------------------------
10 NN for Customer :  32


Customer ID  Similarity Score Rating
2         38828          0.948544      4
3         10661          0.942650      4
4         19381          0.942001      4
5         20901          0.941679      4
6         18049          0.939300      4
7         29803          0.938002      4
8         32828          0.936392      4
9         20750          0.936062      4
10        22819          0.935783      3
11        32065          0.935097      4

----------------------------------
10 NN for Customer :  33


Customer ID  Similarity Score Rating
2         36890          0.939740      3
3          1125          0.939488      3
4          2783          0.938174      3
5         30567          0.930265      3
6         29118          0.925758      3
7          4782          0.925655      3
8          6033          0.925378      2
9         33285          0.921762      3
10        36049          0.920157      3
11        34097          0.919031      2

----------------------------------
10 NN for Customer :  34


Customer ID  Similarity Score Rating
2         26265          0.958907      3
3         33517          0.955095      3
4         32730          0.945025      3
5           825          0.943338      3
6         11343          0.942549      3
7         17766          0.942092      3
8          3480          0.941648      2
9         37531          0.941373      3
10        32905          0.940567      3
11        24768          0.940065      3

----------------------------------
10 NN for Customer :  35


Customer ID  Similarity Score Rating
2         25124          0.935521      3
3         30201          0.934467      2
4         32106          0.933930      2
5          2825          0.932725      3
6          4029          0.932183      3
7          7887          0.929385      3
8           629          0.928248      2
9         35986          0.926892      3
10        37415          0.926600      2
11        35045          0.926440      2

----------------------------------
10 NN for Customer :  36


Customer ID  Similarity Score Rating
2         24217          0.979373      3
3         33294          0.942610      3
4         10486          0.940867      3
5         17785          0.932575      3
6          6374          0.931277      2
7         14502          0.931106      3
8         37522          0.930108      3
9         37692          0.929864      3
10        33456          0.924432      3
11        14664          0.923661      3

----------------------------------
10 NN for Customer :  37


Customer ID  Similarity Score Rating
2           189          0.942801      3
3          3445          0.936213      3
4          2409          0.933562      3
5         10242          0.932968      3
6         34956          0.930480      3
7         31852          0.929065      3
8          7735          0.928203      2
9          6596          0.925417      3
10         9110          0.924012      3
11         3060          0.923105      3

----------------------------------
10 NN for Customer :  38


Customer ID  Similarity Score Rating
2         28695          0.933478      3
3           899          0.920371      3
4         37115          0.916218      2
5         10816          0.913106      3
6         35267          0.911045      3
7         36114          0.907732      3
8         35160          0.907162      3
9         35466          0.904060      3
10        10085          0.901992      3
11         3695          0.901154      3

----------------------------------
10 NN for Customer :  39


Customer ID  Similarity Score Rating
2         14425          0.938034      3
3         34979          0.930302      3
4         25530          0.926674      3
5         34720          0.924116      2
6         33812          0.923922      3
7         39765          0.915980      3
8          9800          0.908617      3
9         20139          0.908135      3
10         3815          0.900640      3
11         2604          0.900630      3

----------------------------------
10 NN for Customer :  40


Customer ID  Similarity Score Rating
2          6926          0.959819      3
3         19147          0.953660      3
4         29340          0.945552      3
5         23209          0.941574      3
6          2437          0.940621      3
7         39107          0.940556      3
8           620          0.938678      3
9         28920          0.938285      3
10        12605          0.937903      3
11         7812          0.937855      3

----------------------------------
10 NN for Customer :  41


Customer ID  Similarity Score Rating
2         11477          0.937826      4
3         28191          0.931394      4
4         18194          0.930732      3
5         28743          0.930160      4
6         18074          0.929413      3
7         14985          0.928435      4
8         21993          0.927850      3
9         10878          0.927115      4
10        15317          0.924111      3
11        22187          0.921660      3

----------------------------------
10 NN for Customer :  42


Customer ID  Similarity Score Rating
2         29451          0.957628      3
3           838          0.956777      3
4         23913          0.955132      3
5         18189          0.953556      3
6         32708          0.951314      3
7         32775          0.949280      3
8          5052          0.947731      2
9          8346          0.942865      3
10        32641          0.942376      3
11        20338          0.941517      3

----------------------------------
10 NN for Customer :  43


Customer ID  Similarity Score Rating
2         15435          0.947671      3
3          6256          0.945252      2
4         35018          0.944433      3
5           182          0.943181      3
6          2423          0.942720      3
7         35407          0.941783      3
8         37734          0.941347      3
9         36887          0.941141      3
10        27788          0.940921      3
11         7070          0.938092      2

----------------------------------
10 NN for Customer :  44


Customer ID  Similarity Score Rating
2         10665          0.939647      3
3           748          0.938333      3
4         12978          0.938202      2
5         36226          0.937940      3
6         28115          0.937751      3
7         21052          0.936178      3
8          4133          0.934322      2
9         21545          0.933725      3
10         3285          0.932265      4
11        25630          0.931933      2

----------------------------------
10 NN for Customer :  45


Customer ID  Similarity Score Rating
2         37575          0.976728      3
3         21660          0.953523      3
4         28866          0.945335      3
5         13252          0.943238      3
6         27775          0.942803      3
7           901          0.942562      3
8         13272          0.941396      3
9         36058          0.938136      3
10          803          0.937395      3
11        32975          0.936668      3

----------------------------------
10 NN for Customer :  46


Customer ID  Similarity Score Rating
2          1626          0.950067      3
3         15115          0.941860      3
4         40704          0.940693      3
5         38028          0.939905      3
6          8658          0.937739      3
7           630          0.936404      2
8         42704          0.936152      3
9         15022          0.933434      3
10        24108          0.933140      3
11        35218          0.932783      3

----------------------------------
10 NN for Customer :  47


Customer ID  Similarity Score Rating
2          3292          0.946503      2
3         34864          0.927473      2
4          3583          0.926334      2
5          1327          0.926249      2
6         33436          0.924858      2
7         25445          0.923783      2
8         33756          0.920580      2
9         36362          0.919929      2
10         5701          0.919469      2
11        15132          0.918314      1

----------------------------------
10 NN for Customer :  48


Customer ID  Similarity Score Rating
2         15971          0.911205      2
3         34537          0.899792      3
4         16935          0.893514      3
5         17586          0.881529      3
6         25651          0.881168      3
7         29163          0.878484      3
8         15224          0.876908      3
9         27214          0.875713      3
10        27489          0.872171      3
11        12390          0.871361      3

----------------------------------
10 NN for Customer :  49


Customer ID  Similarity Score Rating
2         41203          0.955524      3
3          7355          0.942810      3
4         39790          0.941534      3
5         19033          0.938969      4
6         27828          0.938584      3
7           747          0.938221      3
8         10060          0.936804      3
9         33667          0.935400      3
10        20338          0.933957      3
11        32641          0.932703      3

----------------------------------
10 NN for Customer :  50


Customer ID  Similarity Score Rating
2          7958          0.944307      3
3         33871          0.930270      3
4         41573          0.928048      3
5         27472          0.924487      3
6          7287          0.923764      3
7         17936          0.917568      2
8           912          0.916046      2
9          9702          0.915160      3
10        24817          0.913946      3
11        26418          0.909517      3

###  4.2) Prediction of the Customer Rating 
Based only on the 10 most similar customers computed in the previous step, predict the customer rating rank using:
* The average rating rank of the 10 most similar customers (rounded to the nearest integer).
* The weighted average rating rank of the 10 most similar customers (rounded to the nearest integer).

$$
Weighted~average~rating~rank = \frac{\sum_{i=1}^{10}[similarity(i)*rank(rating(i))]}{\sum_{i=1}^{10}similarity(i)}
$$

Where:
* `rating(i)` = the rating of the i-th nearest neighbor (i=1 for the most similar customer)
* `similarity(i)` = the similarity of the i-th nearest neighbor with the given customer



##### METHOD 1 : Prediction of customer Rating using the `average rating` of the 10 most similar customers 

* In order to compute the average rating of the 10 most similar customers we have constructed a `for` loop. The loop runs through the datasets which are stored in the `global` variable and computes the average `Rating` of every dataframe ,  `rounds it` to the nearest integer and `stores` them into a list called __`average_rating`.__

In [29]:
average_rating=[]
for i in range (1,51):
    score_rating = globals()[f"df_{i}"]['Rating'].mean().round()
    average_rating.append(score_rating)

* After the computation of the average rating we tranformed the list into a pandas DataFrame.
   * We renamed the index of the DataFrame into `Customer ID` and the column of the scores into `Predicted Average Rating`.
   * Also because our customer ids need to start from 1 we had to increase the index by 1.
* Finally , The  `Average Rating` dataframe contains the `Predicted Average Rating` of the 10 most similar customers for our first 50 customers.

In [30]:
average_rating = pd.DataFrame(average_rating).rename_axis(index="Customer ID").set_axis(['Predicted Average Rating'],axis=1)
average_rating.index = average_rating.index+1
average_rating.reset_index().head(10)

Customer ID  Predicted Average Rating
0            1                       2.0
1            2                       3.0
2            3                       2.0
3            4                       2.0
4            5                       3.0
5            6                       3.0
6            7                       3.0
7            8                       3.0
8            9                       3.0
9           10                       3.0

* The final step is to merge the initial dataset `banks` with the `average rating` on the `Customer ID` column. After dropping the unnecessary columns we managed to have side by side the `True Rating` and the `Predicted Rating` of the first 50 customers.

In [31]:
average_rating = pd.merge(banks, average_rating, how='inner', on = 'Customer ID')
average_rating=average_rating.drop(['Age','Job','Marital','Education','Default','Balance','Housing','Loan','Products'],axis=1)

##### Final Dataframe with the `Rating` and the `Predicted Average Rating`

* In order not to have misunderstandings about which index corresponds to every customer we have made the index start from 1 to be the same.

In [32]:
average_rating.index=average_rating.index+1
average_rating

Customer ID  Rating  Predicted Average Rating
1             1       1                       2.0
2             2       3                       3.0
3             3       2                       2.0
4             4       2                       2.0
5             5       3                       3.0
6             6       3                       3.0
7             7       3                       3.0
8             8       3                       3.0
9             9       3                       3.0
10           10       3                       3.0
11           11       3                       3.0
12           12       3                       3.0
13           13       3                       3.0
14           14       3                       3.0
15           15       2                       2.0
16           16       3                       3.0
17           17       2                       2.0
18           18       3                       3.0
19           19       2                       2.0
20           20       3                       3.0
21           21       3                       3.0
22           22       2                       2.0
23           23       2                       3.0
24           24       2                       2.0
25           25       3                       3.0
26           26       3                       3.0
27           27       2                       2.0
28           28       2                       3.0
29           29       4                       3.0
30           30       3                       3.0
31           31       2                       3.0
32           32       3                       4.0
33           33       3                       3.0
34           34       2                       3.0
35           35       3                       2.0
36           36       3                       3.0
37           37       3                       3.0
38           38       2                       3.0
39           39       2                       3.0
40           40       2                       3.0
41           41       3                       4.0
42           42       2                       3.0
43           43       3                       3.0
44           44       2                       3.0
45           45       2                       3.0
46           46       2                       3.0
47           47       1                       2.0
48           48       3                       3.0
49           49       3                       3.0
50           50       3                       3.0

##### METHOD 2 : Prediction of customer Rating using the __`Weighted Average Rating`__ of the 10 most similar customers

* In order to compute the `weighted average rating` of the 10 most similar customers we have constructed a `for` loop. The loop runs through the datasets which are stored in the `global` variable and computes the weighted average `Rating` of every dataframe ,  rounds it to the nearest integer and stores them into a list called __`average_rating_2`.__
* The `j` runs from 2 to 11 because we exclude the first row which contains the score of the customer himself. It would always be equal with 1.
* The function that computes the weighted average rating has been mentioned in the beginning of the `paragraph 4.2`.

In [33]:
%%time
average_rating_2=[]
for i in range(1,51):
    score_2 = 0
    for j in range(2,12):
        score_2 = score_2 + ( globals()[f"df_{i}"]['Similarity Score'][j] ) * globals()[f"df_{i}"]['Rating'][j] 
    score_2 = (score_2 / sum( globals()[f"df_{i}"]['Similarity Score'])).round()
    average_rating_2.append(score_2)

Wall time: 7 ms


* After the computation of the weighted average rating we tranformed the list into a pandas DataFrame.
   * We renamed the index of the DataFrame into `Customer ID` and the column of the scores into `Predicted Weighted Average Rating`.
   * Also because our customer ids need to start from 1 we had to increase the index by 1.
* In more details the `Average Rating 2` dataframe contains the `Predicted Weighted Average Rating` of the 10 most similar customers for our first 50 customers.

In [34]:
average_rating_2=pd.DataFrame(average_rating_2).rename_axis(index="Customer ID").set_axis(['Predicted Weighted Average Rating'],axis=1)
average_rating_2.index = average_rating_2.index+1
average_rating_2.reset_index().head(5)

Customer ID  Predicted Weighted Average Rating
0            1                                1.0
1            2                                3.0
2            3                                2.0
3            4                                2.0
4            5                                3.0

* The final step is to merge the initial dataset `banks` with the `average rating_2` on the `Customer ID` column. After dropping the unnecessary columns we managed to have side by side the `True Rating` and the `Predicted Rating` of the first 50 customers. 

In [35]:
average_rating_2 = pd.merge(banks, average_rating_2, how='inner', on = 'Customer ID')
average_rating_2 = average_rating_2.drop(['Age','Job','Marital','Education','Default','Balance','Housing','Loan','Products'],axis=1)

##### Final Dataframe with the `Rating` and the `Predicted  Weighted Average Rating`

* In order not to have misunderstandings about which index corresponds to every customer we have made the index start from 1 to have the same.

In [36]:
average_rating_2.index =average_rating_2.index+1
average_rating_2

Customer ID  Rating  Predicted Weighted Average Rating
1             1       1                                1.0
2             2       3                                3.0
3             3       2                                2.0
4             4       2                                2.0
5             5       3                                3.0
6             6       3                                3.0
7             7       3                                3.0
8             8       3                                3.0
9             9       3                                3.0
10           10       3                                3.0
11           11       3                                3.0
12           12       3                                3.0
13           13       3                                3.0
14           14       3                                3.0
15           15       2                                2.0
16           16       3                                3.0
17           17       2                                2.0
18           18       3                                3.0
19           19       2                                2.0
20           20       3                                3.0
21           21       3                                3.0
22           22       2                                2.0
23           23       2                                3.0
24           24       2                                2.0
25           25       3                                3.0
26           26       3                                3.0
27           27       2                                2.0
28           28       2                                3.0
29           29       4                                3.0
30           30       3                                3.0
31           31       2                                3.0
32           32       3                                4.0
33           33       3                                3.0
34           34       2                                3.0
35           35       3                                3.0
36           36       3                                3.0
37           37       3                                3.0
38           38       2                                3.0
39           39       2                                3.0
40           40       2                                3.0
41           41       3                                4.0
42           42       2                                3.0
43           43       3                                3.0
44           44       2                                3.0
45           45       2                                3.0
46           46       2                                3.0
47           47       1                                2.0
48           48       3                                3.0
49           49       3                                3.0
50           50       3                                3.0

### 4.3 ) Evaluation of the Classification Algorithm
For the evaluation of your classification algorithm we will use the 50 first records of the bank dataset and predict the rating for them. Then, for all n=50 records calculate the Mean Prediction Error for both prediction methods.

$$
Mean~Prediction~Error = \frac{\sum_{i=1}^{n}|rank(Predicted~rating(i))-rank(True~rating(i))|}{n}
$$


##### Mean Prediction Error for Method 1 ( Predicted Average Rating )

In [37]:
MPE= sum(abs(average_rating['Predicted Average Rating'] - average_rating['Rating'] ) ) / 50
print("The Mean Prediction Error of the 1st method :",f"{MPE:.0%}")

The Mean Prediction Error of the 1st method : 34%


##### Mean Prediction Error for  Method 2 ( Predicted  Weighted Average Rating )

In [38]:
MPE_2 = sum(abs(average_rating_2['Predicted Weighted Average Rating'] - average_rating_2['Rating'] )) / 50
print("The Mean Prediction Error of the 2nd method :",f"{MPE_2:.0%}")

The Mean Prediction Error of the 2nd method : 30%


##### Final conclusion 

* The Prediction Error is the absolute value of the difference between the predicted value and the actual value. MPE tells us how big of an error we can expect from the prediction on average. The lower value the model has the better is its  performance
* As we can see from the outputs above , the first method results in a bigger Mean Prediction Error than the second method which means that our predictive algorithms have an error percentage of 34% and 30% respectively.

* Model 1 : `MPE=0.34` implies that, on average, the Prediction's distance from the true value is 0.34 or 34%.
* Model 2 : `MPE=0.30` implies that, on average, the Prediction's distance from the true value is 0.3 or 30%.

Finally , we can say that the best model for classification is the `Model 2` which is using for the Prediction of the customers `Rating the Weighted Average Rating`.